### DISCUSSION TITLE
Why Next 13.4 wants me to add "use-client" to a custom hook?

### DISCUSSION THREAD

**Discussion by Maverick17user** [vercel/next.js#50549] [App Router] [answered]
Hi! I'm trying to migrate my code-base from pages to app dir routing.

Question: why next.js treat my custom `useOpen` hook as component and want me to add "use client" there?
I thought we need define "use client" only in .tsx/.jsx files.

<img width="1008" alt="image" src="https://github.com/vercel/next.js/assets/30416868/84f4973c-9f5f-44f2-8659-9784190414a9">

Hook code:

<img width="760" alt="image" src="https://github.com/vercel/next.js/assets/30416868/2554499b-6a0a-46df-9f82-da169f29ca58">

This hook uses in 3 "use-client" components:

<img width="333" alt="image" src="https://github.com/vercel/next.js/assets/30416868/280c6a3c-7bc1-40ac-a8ca-9285b1b874ac">

Why this might be happen?

**Comment by paqstd-dev** [comment]
Hi, according to the [documentation](https://nextjs.org/docs/getting-started/react-essentials#when-to-use-server-and-client-components), all hooks should only work on the client. At the specified link, some patterns are also written that can help in development.

**Reply by EliyahuCohen** [reply]
make sure that you are not inside of the app folder and than the need for the "use client" is not required

**Comment by icyJoseph** [comment]
Interactivity is only available client side, so a module that wants to do any sort of interactivity has to be a `use client`. Note that I said module. 

- https://nextjs.org/docs/getting-started/react-essentials#when-to-use-server-and-client-components

![Screenshot 2023-05-31 at 09 17 32](https://github.com/vercel/next.js/assets/21013447/364e5933-1830-49ea-ba59-665945130577)

You have to think in terms of module graph, rather than component graph.

> "use client" sits between server-only and client code. It's placed at the top of a file, above imports, to define the cut-off point where it crosses the boundary from the server-only to the client part. Once "use client" is defined in a file, all other modules imported into it, including child components, are considered part of the client bundle.

**Comment by Maverick17user** [comment]
I have huge amount of such modules, and adding to all of them "use client" seems ... bad 💀

**Reply by icyJoseph** [reply]
but then the thing is that you are importing this into a Server Component. Right? You can skip it at module level, if the importing parent, has "use client".

The problem is that components are Server Components by default. You have to use the "use client" directive to mark them as client components. 

As the last bit on my previous answer says:

> Once "use client" is defined in a file, all other modules imported into it, including child components, are considered part of the client bundle.

**Reply by icyJoseph** [reply]
For what is worth, I think I was very unclear in my other answer.

**Reply by Maverick17user** [reply]
> but then the thing is that you are importing this into a Server Component. Right?

In my case I import this hook into a client components only and Next.js still want me to add "use client" to the module with this hook.

**Reply by icyJoseph** [reply]
Really? I'll give it a go. Do you export other things from the module? That are maybe used in a Server Component?

**Reply by Maverick17user** [reply]
<details>

<summary>Yeah ...
</summary>

<img width="976" alt="image" src="https://github.com/vercel/next.js/assets/30416868/98b2f3b7-843a-4196-ae0c-ebedb0703dcf">


</details>

<details>

<summary>But I have this module with hook outside my app folder. Does it influence to the end result or not?

</summary>

<img width="440" alt="image" src="https://github.com/vercel/next.js/assets/30416868/00ec4807-1df1-4a56-bfbc-ef202e1caf2b">


</details>

<details>

<summary>Components, which uses the hook from the module, which Next wants to start with "use client" (all components locate in src/components)
</summary>

### 1
![image](https://github.com/vercel/next.js/assets/30416868/7618f28a-b026-4b57-989a-cb1e99a0587b)

### 2
![image](https://github.com/vercel/next.js/assets/30416868/e03a18b4-ca15-4f92-a250-d10a61a0d5e6)

### 3
![image](https://github.com/vercel/next.js/assets/30416868/b5be5bfe-da3b-4482-8fa5-f288afbeefef)

</details>


I'm not sure, does it some next.js bug, or I do smth. wrong?

**Reply by Maverick17user** [reply]
<details>

<summary>Another interesting moment - I face this error at "/" page, but that page doesn't use any of these components and that hook.
</summary>


app root page.tsx
<img width="825" alt="image" src="https://github.com/vercel/next.js/assets/30416868/7bfe0160-0cef-48bf-9cb8-3446a1d8de14">

app root layout.tsx
<img width="1059" alt="image" src="https://github.com/vercel/next.js/assets/30416868/68fa5ee2-48fa-4275-96df-6e5b088d7fff">

</details>

**Reply by icyJoseph** [reply, marked as answer]
Yeah but you have `import { colors } "@core";` which I guess touches a file `core/index.ts`, which in turn probably, imports `useOpen`, somehow, right?

So my question: Do you export other things from the module?  is kind of true, you are importing this hook into page, albeit not directly, but through a barrel file of sorts, right?

**Reply by icyJoseph** [reply]
Seeing the import chain as a graph is important, otherwise just importing can often to lead to unpredictable, or non-intuitive issues.

In this commit for example https://github.com/icyJoseph/next-js-react-gbg/commit/3610a9dc32ba29f202c3b31bd44e7f67cb334ee0 I fixed a problem where importing a simple `string` constant, from a file that contained a function that used `crypto`, lead to browserify crypto being included on my client bundle. That's about 200kB I surely did not want... 

In your case I think, you import colors from a file that in turn re-exports imports from other files, one of them ends up in a component that uses `client component hooks`, and since the top level importer is a Server Component, it breaks.

**Reply by Maverick17user** [reply]
@icyJoseph, I suppose you're right! 

I've changed my code like this and it fixed my problem.
<img width="412" alt="image" src="https://github.com/vercel/next.js/assets/30416868/51dbe1c7-1d32-442a-abc8-8ac9b25e32bf">

A @core dir in my project has this kind of structure
<img width="634" alt="image" src="https://github.com/vercel/next.js/assets/30416868/d37a8880-83ec-4bce-8cca-ba880b4d65be">

And I do kind of shortcut in my tsconfig file to make import paths shorter
<img width="325" alt="image" src="https://github.com/vercel/next.js/assets/30416868/3478bd12-b89e-4fad-93b3-cf610217a478">

I've also did like that in me core/index.ts and error also been fixed
<img width="289" alt="image" src="https://github.com/vercel/next.js/assets/30416868/3f90c539-85d3-47a1-942c-5829505ba655">

So, i have my reusable react components and other stuff in my core folder. 
Do I need now to have kinda coreClient & coreServer folders to avoid such conflicts?

**Reply by Maverick17user** [reply]
So, both useOpen hook & my coors constant imports from the same index.ts

**Reply by icyJoseph** [reply]
Yeah, I suppose the root issue here is that certain tooling has created an expectation with regards to tree shaking and code splitting, which in turn lead to the so-called barrel file pattern. Where modules are re-exported by an entry point.

> We are also kind of blindly trusting webpack through next.js, plus the plugins and actual implementation, so the less we assume and take for granted the better.

This was already problematic when testing. For example Jest, would enter the barrel file and parse everything, absolutely, and then take the one little thing you actually want to test. When barrel files mixed code from various environments, you ended up with failed tests, och absurd mocks just get going. Say testing within `jsdom` and hitting a Node API.

Now that server components present the module as the boundary, barrel files are even more strange. They could break environment isolation, and are somewhat redundant step since the modules they re-export, are already public.

Obviously if the re-exported members of a barrel file are all agnostic to the environment, or belong to the same environment, then we are good.

So yeah, you are probably better off separating your code by environment. 

Some libraries export conditionally too, but that's a topic I won't go into, for various reasons.

**Reply by icyJoseph** [reply]
Also please don't forget to mark as answered if you got what you needed :) keep on asking! I find these problems in the migration from pages to app router, very interesting.

**Reply by Maverick17user** [reply]
Thx a lot, after some refactoring I made entire app to work via app-folder -oriented approach. Much appreciated for your help :)

**Comment by Kosa103** [comment]
I've found the solution. For me it happened because of barrel files. These are the index.ts files thats sole purpose is to re-export other exports nested deeper in the folder tree. They seem handy to make imports shorter but when you use the short import it actually loads everything from that file. So if you re-export constants, helpers and hooks from a barrel file and you import one of the constants in a server component it will try to load all your hooks too! And of course it fails.

I hope this helped!
